### ISSUE TITLE
False unused_assignments warnings

### ISSUE THREAD

**Issue by terrarier2111** [rust-lang/rust-clippy#9918]
### Summary

When trying to compile the provided code a false-positive `unused_assignments` warning gets emitted although there shouldn't be any warning.

### Lint Name

unused_assignments

### Reproducer

I tried this code:

```rust
loop {
                        let mut curr = 0;
                        match atomic.compare_exchange(curr, Self::UPDATE, Ordering::AcqRel, Ordering::Relaxed) {
                            Ok(_) => {
                                // do stuff!
                                break;
                            }
                            Err(_) => {
                                if atomic.fetch_or(Self::OTHER_UPDATE, Ordering::AcqRel) == 0 {
                                    curr = Self::OTHER_UPDATE;
                                    continue;
                                }
                                break;
                            }
                        }
                    }
```

I saw this happen:

```
warning: value assigned to `curr` is never read
   --> src/file.rs:471:37
    |
471 | ...                   curr = Self::OTHER_UPDATE;
    |                       ^^^^
    |
    = help: maybe it is overwritten before being read?
    = note: `#[warn(unused_assignments)]` on by default
```

I expected to see this happen:

No warnings

### Version

```text
rustc 1.66.0-nightly (9565dfeb4 2022-10-28)
binary: rustc
commit-hash: 9565dfeb4e6225177bbe78f18cd48a7982f34401
commit-date: 2022-10-28
host: x86_64-unknown-linux-gnu
release: 1.66.0-nightly
LLVM version: 15.0.2
```

**Comment by matthiaskrgr** [comment]
This looks like a rustc warning and not a clippy warning ^^

Anyway I think rustc is correct here
0) loop starts
1) we initialize  `let mut curr = 0;`
2) `0` is passed to the `compare_exchange` of the match.
3) if we reach `Err..` we might set `curr = Self::OTHER_UPDATE;` and continue into the next loop iteration 
=> in the next iteration, `curr` is reset to 0 again.

So IMO rust is correct here and the `curr = Self::OTHER_UPDATE;` is indeed never read before it is overwritten by a re-initialisation. at line 2

**Comment by y21** [comment]
Closing as there's indeed no bug here.
